[Videoplayer] Audio Channel Layout Infolabels

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

This change creates two infolabels `ListItem.AudioChannelLayout` and `VideoPlayer.AudioChannelLayout` mentioned in comments of #26626 for ffmpeg, dvd and inputstream sources.

A default layout based on the channel count is provided when the actual layout cannot be extracted (ex. Inputstream Adaptive). The channel layout part of CodecDesc will also be synchronized to have the same values.

![image](https://github.com/user-attachments/assets/232bec59-b30b-43c5-8248-81a7400f108a)

![image](https://github.com/user-attachments/assets/6c13e7ae-fe76-4751-95ab-85b2a5c62748)

![image](https://github.com/user-attachments/assets/b6aba5b4-f521-4ca0-b1a9-ece0cee4b71e)

![image](https://github.com/user-attachments/assets/e215c621-0e67-41b2-89f3-a63ab696774f)

![image](https://github.com/user-attachments/assets/98a6d7f1-a07e-4c3f-9dfe-bb9a4773a384)

![image](https://github.com/user-attachments/assets/7e04ee7d-598e-4eb3-b9e7-5a4eb561462d)

xx channels could be removed here, either by using the same format as stream selection or by using codecdesc alone, which now always provides channel information.
![image](https://github.com/user-attachments/assets/8e55aa6e-4fe2-48ea-b65a-eb1f8458f6b3)

Could replace channels here:
![image](https://github.com/user-attachments/assets/7d08b0c2-c280-42e4-81ff-7f1c2a99fb4b)

No change planned here, these are the actual channels open for output:
![image](https://github.com/user-attachments/assets/3b9d550d-a106-4e2d-aa46-b4c6d775565e)

Input stream adaptive receives the default layouts:
![image](https://github.com/user-attachments/assets/c99a271c-5979-4d9d-a3d1-440ded95a301)
(channel layout is meaningless for Atmos)

Integration to standard stream details functions had to be implemented at the same time: extraction from new files, at the start/during playback, read/export to .nfo and video db.
~video db: new text field streamdetails.strAudioChannelLayout
.nfo: new `<audiochannellayout>` tag inside `<fileinfo><streamdetails><audio>`~
edit for mask approach:
video db field: streamdetails.strAudioChannelMask, type bigint (64bits)
.nfo: new `<channelmask>` tag inside `<fileinfo><streamdetails><audio>`

Based on feedback, ffmpeg default layouts for a given channel count are overriden in a few cases to match with historical conversions in Estuary.
See below some data on ffmpeg vs current Estuary.

Default layouts:
| channels | Current Estuary | ffmpeg default layout name (when different) | Localizable |
|:---|:---|:---|:---|
| 0 | 0.0 | 0 channels | yes - strings.po 10127 |
| 1 | 1.0 | mono |  yes - strings.po 37003 |
| 2 | 2.0 | stereo | yes - strings.po 37004 |
| 3 | 2.1 | |
| 4 | 4.0 | |
| 5 | 4.1 | 5.0 | 
| 6 | 5.1 | |
| 7 | 6.1 | |
| 8 | 7.1 | |
| 9 | | 9 channels | yes - strings.po 10127 |
| 10 | 9.1 | 5.1.4 |  
| 11 | | 11 channels | yes - strings.po 10127 |
| 12 | 7.1.4 | 
| 13 | | 13 channels | yes - strings.po 10127 |
| 14 | | 9.1.4 |
| 15 | | 15 channels | yes - strings.po 10127 |
| 16 | 9.1.6 | hexadecagonal | yes |

Some additional layouts (same channel count as above, but different channels. detected by ffmpeg)
| channels | ffmpeg layout name | Localizable | Comment |
|:---|:---|:---|:---|
| 6 | 5.1(side) | yes | the most common type of 5.1 |
| 4 | quad | yes |
...

note: The Estuary change is just a quick hack meant to show the new infolabels for videos instead of the skin-based lookups and will not be merged.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Provide accurate and consistent layout information when available and replace the heuristics coded in each skin.
For example with 4 channels it's not possible to guess if the layout is 3.1, 4.0, quad or quad(side)

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Database tests:
With empty database to test automatic extraction of stream details of files
With database of previous version to test the upgrade function

Display tests:
Media flags displayed on home page, library/videos views, during playback the stream selection and OSD Audio Settings.
- local file
- test stream for input stream adaptive 

.nfo import/export for movie

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

More accurate/consistent indication of audio channel layout.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [X] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[VideoPlayer] Use default mask based on channel count for unknown masks

CrystalP committed:
[VideoPlayer] Provide default audio channel layout for DVD

CrystalP committed:
[VideoPlayer] Provide guidance for future population of actual channel layout for inputstream adaptative

CrystalP committed:
[Estuary] incomplete modification to show core-provided audio channel layout for video

chewitt commented:
I'm trying to think up a good reason for adding text like `(side)` and `hexadecagonal` into the GUI and I'm not coming up with anything. Those terms might mean something internal to ffmpeg but for normal users the existing numerial format Estuary uses is visually simple and the extra text adds clutter.

CastagnaIT commented:
> I'm trying to think up a good reason for adding text like (side) and hexadecagonal into the GUI 

no i had time to look at anything from the PR just screenshots but
i agree some ffmpeg texts are weird...
the most correct thing would be to have our own textual layout description without depending on the texts descriptions provided by ffmpeg,
ffmpeg should just tell us the exact layout format, and so “translate” the format with an our own text
assuming it is possible to do something similar...

CastagnaIT commented:
maybe likely having a `float `channel number instead of `int` allows us to differentiate a bit better
2.1 vs 3
3.1 vs 4
4.0 vs 4
4.1 vs 5
5.0 vs 5
i found this list 
https://github.com/FFmpeg/FFmpeg/blob/ec3f3457fdfdf49b64e57c1e6c8cefd0603bc309/libavutil/channel_layout.h#L394-L435
I don't know if it can helps

jjd-uk commented:
Will need to check this out against my large library of files with differing audio formats/layouts.

> maybe likely having a `float `channel number instead of `int` allows us to differentiate a bit better 2.1 vs 3 3.1 vs 4 4.0 vs 4 4.1 vs 5 5.0 vs 5 i found this list https://github.com/FFmpeg/FFmpeg/blob/ec3f3457fdfdf49b64e57c1e6c8cefd0603bc309/libavutil/channel_layout.h#L394-L435 I don't know if it can helps

If that is differnet to what is used in this PR, then that looks to be a almost complete list that would seem to cover the layouts that are most difficut to covert from a simple integer number of channels only. 

For example is 10 channels 7.1.2 or 5.1.4 would look to be covered by:

1. AV_CHANNEL_LAYOUT_7POINT1POINT2
2. AV_CHANNEL_LAYOUT_5POINT1POINT4_BACK

Hitcher commented:
Also see: https://ffmpeg.org/doxygen/7.0/channel__layout_8h.html

ie `#define 	AV_CHANNEL_LAYOUT_7POINT1POINT2   AV_CHANNEL_LAYOUT_MASK(10, AV_CH_LAYOUT_7POINT1POINT2)`

jjd-uk commented:
As Atmos are some of the most difficult to get right, if anyone requires test files then they can be sourced direct from Dolby.

https://download.dolby.com/us/en/test-tones/dolby-test-tones_5_1_2.mp4
https://download.dolby.com/us/en/test-tones/dolby-test-tones_5_1_4.mp4
https://download.dolby.com/us/en/test-tones/dolby-test-tones_7_1_2.mp4
https://download.dolby.com/us/en/test-tones/dolby-test-tones_7_1_4.mp4
https://download.dolby.com/us/en/test-tones/dolby-test-tones_9_1_6.mp4

CrystalP commented:
The ffmpeg layout names are currently used as is in the CodecDesc infolabel (appears in OSD audio settings in Estuary).
A few of the names seem more music related (mono, stereo, quad). I don't particularly care about that.
The (side) (front) etc guarantee uniqueness and allow going back and forth between text and internal representation losslessly.

If we don't care about uniqueness, we can use the x.y.z format. No problem to create from an ffmpeg layout structure.
With x - speakers on the normal plane, y - LFE, z - overhead speakers.
It means that 5.1 with back or side surrounds will look identical for example. Bonus: no localization needed with that form.

If we care about keeping detailed and unambiguous layouts, the channels mask in binary form could be used/stored internally and converted to x.y.z form by the infolabels when used by a skin. It won't be as clear as a x.y.z text in the db or nfo.
For example, the AV_CH_LAYOUT_5POINT1, text 5.1(side), is 1551 as integer representing the binary mask.

jjd-uk commented:
For GUI I'd say x.y.z format is fine, since I'd assume the log would still show the unique value e.g. 5.1(side) if someone wants to dig deeper, as it seem to me that it would likely only be necessary if something unexpected is happening.

Hitcher commented:
x.y.z sounds good to me.

Hitcher reviewed (approved):
Everything seems to be working as expected on the files I tested.

Thanks for this addition, I've wanted it for some time now.

CastagnaIT reviewed (changes requested):
I don't currently find this PR so accurate,
it introduces a new field about layout and it still tied to internal values in ffmpeg

although the old “codecDesc” uses it IMO is not correct continue in this way,
we need our own kodi definitions, and so apply it also to “codecDesc” field

> If that is differnet to what is used in this PR, then that looks to be a almost complete list that would seem to cover the layouts that are most difficut to covert from a simple integer number of channels only.

in fact kodi uses an integer which is (at today) inappropriate, or we change data type
or think of an alternative to still use integer values which will probably turn out to be non-human readable
but if we have to make changes to the best of our ability, let's consider the various possibilities

I think for this PR should handle our own list of layout descriptions and is also applied on codecDesc

pro:
-not bound to ffmpeg internal strings that cannot be controlled by us
-we have layout strings defined by us, that will not change over time (ffmpeg can change their string for their needs)
-we can avoid quirky values like hexadecagonal etc...

CrystalP commented:
@jjd-uk regarding atmos tests
ffmpeg reports channels correctly for non-atmos truehd or eac3, including the dependent channel additions/substitutions - I tested a 7.1 eac3 successfully.

With atmos the channel count loses its meaning and I'm not sure a channel count or layout makes senses today or even later once ffmpeg is eventually able to decode JOC.

The atmos test tones samples technically require only one channel of audio in the file (so could be 1.0), with object positioning data to move that channel exactly to each speaker of a 7.1.4 setup (for example) in turn.
For backwards compatibility with non-atmos decoders I suppose, a downmix to 5.1 is included and seen by ffmpeg.

CrystalP committed:
[VideoPlayer] Add audio channels mask to videoinfotag / player process info / stream selection

CrystalP committed:
[VideoPlayer] Add audio channels layout infolabels + supporting utils functions

CrystalP commented:
Updated the PR using feedback above and overrode some of ffmpeg's default layouts to match Estuary's current conversions (easy to adjust). + rebased on more recent master.

The layout is stored internally as a bit mask of channels to retain the full details and converted to X.Y.Z form by the infolabels code.
The same values are displayed everywhere: listitem, videoplayer, stream selection, audio osd list of tracks (ie codecdesc)

- existing values in database, inputstream adaptive, dvd: default channels mask & layout from the channel count (CStreamUtils::GetDefaultMask)
- everything else: use the mask return by the ffmpeg demuxer. ffmpeg returns a blank mask for some streams (I found a 1 channel PCM example - MediaInfo also seems to think there is no channel layout for that file).

For now, in addition to X.Y.Z the labels also return "X channels" when the ffmpeg demuxer doesn't provide a proper mask.

- database version note: this PR bumps the db but a recent PVR PR bumped as well. If you ran the previous iteration of this PR, you must rollback the db to 134 for a proper conversion of existing stream details. If you didn't run the PR, then nothing special to do.

Hitcher commented:
Just tested this quickly but most of media flags in the library now show 2.0 when they're 5.1. The layout is correct when checking the stream details or from the player process dialog.

CrystalP commented:
@Hitcher strange, this works fine here and  indicates a problem with the migration of the existing stream details data.
I looked closer and didn't find a problem that would explain what you saw, but found that I was not using the correct db column data type and fixed it in the force-push just now, please grab the latest commits before you try again.

Did you run the previous version of the PR? Did you go back to db 134 before starting the new version?

Could you share a screenshot of a few rows of the streamdetails table for files with issues, showing the  iAudioChannels and strAudioChannelMask colums.

Hitcher commented:
> Did you run the previous version of the PR? Did you go back to db 134 before starting the new version?

My mistake, I had v135 and v136. Removed them up both and used v134 and all is good.

CastagnaIT commented:
better solution
but correct me if im wrong, from what i understand you are saving to db ffmpeg flags (as "mask")
so you are always tied to possible breakups...

if one day we bump ffmpeg version and upstream has changed some flag values without that we know it, whole database values ​​will be corrupted (i remember that in the past it was discussed that kodi can be built with different versions of ffmpeg)

so why not save to db already translated layout value as a string instead of "mask" byte? you would not be tied to the flags

CastagnaIT commented on xbmc/utils/StreamUtils.cpp:
```diff
@@ -96,3 +100,64 @@ std::string StreamUtils::GetCodecName(int codecId, int profile)
+  if (mask)
+    return GetLayoutXYZ(mask);
+
+  //! @todo return something or nothing in case of unknown layout?
+  std::string layout = std::to_string(channels);
+  layout.append(" ");
+  layout.append(g_localizeStrings.Get(10127)); // "channels"
+  return layout;
```
IMO you should avoid use language localizations here
and leave the task to skins/addons etc...
otherwise the value will result impossible to be mapped

CrystalP replied:
We could mark the message po.strings as used by skins - I think it's of general interest and defining it per skin/addon would be more effort overall.

Whether to display a xx channels in case of no data or return a blank string and let the skin handle - that is the fundamental question here. I'd let skinners answer.

unknown layout: as things currently stand, when ffmpeg demuxer doesn't find a layout in the video, or (every other flow), we don't have a default layout for a channel count.

CastagnaIT replied:
since already exists a numeric channels
i think that mixing layouts and custom form of channels layout (with also localization) can be avoided

skins and addons have the ability to handle this
then No layout == empty string,
skins can fallback to use integer channel number and also use localization when the layout is empty

if on skin translations string dont exist a relative string for "channels" (10127) then must be add a new one on skin translations
you have not to mix kodi core translations with skin

CrystalP replied:
Added a commit to leave the infolabel blank (to be squashed later)

CrystalP commented:
> better solution but correct me if im wrong, from what i understand you are saving to db ffmpeg flags (as "mask") so you are always tied to possible breakups...
> 
> if one day we bump ffmpeg version and upstream has changed some flag values without that we know it, whole database values ​​will be corrupted (i remember that in the past it was discussed that kodi can be built with different versions of ffmpeg)
> 
> so why not save to db already translated(i mean the result x.y.z) layout value as a string instead of "mask" byte? you would not be tied to the flags

I'm not against storing the formatted string instead of the mask. The consequence is losing some accuracy that won't be recoverable later on. As long as it's understood and accepted I'm fine.
There would actually be a slight performance improvement as the stored text would be ready to display (as long as there is no localization in any of the layout values)

ffmpeg version: not worried. From ffmpeg's own documentation about versioning and compatibility:
```
FFmpeg guarantees backward API and ABI compatibility for each library as long as its major version number is unchanged. This means that no public symbols will be removed or renamed. Types and names of the public struct members and values of public macros and enums will remain the same (unless they were explicitly declared as not part of the public API). Documented behavior will not change.
```

If we still wanted to "own" the API, AE could be extended with additional speaker locations I suppose but is it a public API?

CrystalP committed:
Blank infolabel when a mask is not provided/cannot be guessed.
To be squashed with the commit #2 that create StreamUtils::GetLayout

CrystalP commented:
Added a commit to make the infolabels blank for an unknown layout based on comment. The skin will be responsible to provide a fallback text such as "x channels" instead.
The force-push is because the use of the "channels" message by DVDInputStreamNavigator was not removed in the correct commit. Everything else unchanged.

@CastagnaIT do you have further comments? (see as well my last answer about storage of binary mask vs x.y.z string)

CastagnaIT commented:
regarding to save to db the mask int VS layout x.y.z. string, i have no further questions, if everyone is okay with that
i will do a another PR review tonight or tomorrow

jjd-uk commented:
Haven't had time for a detailed look at this yet. Will make time in the coming week.

CrystalP commented:
For comparison I created #26747 which stores and uses internally the x.y.z strings instead of the channels mask.

One could say that using the channels masks is over-engineering for a future problem that may never happen.

Delete the videodb 136 when going back and forth since both code versions don't create the same database columns.

CastagnaIT commented on xbmc/utils/StreamDetails.cpp:
```diff
@@ -475,6 +480,16 @@ int CStreamDetails::GetAudioChannels(int idx) const
+      dynamic_cast<const CStreamDetailAudio*>(GetNthStream(CStreamDetail::AUDIO, idx));
+  if (item)
+    return item->m_channelMask;
+  else
```
"else" condition can be removed

CastagnaIT commented on xbmc/utils/StreamUtils.cpp:
```diff
@@ -96,3 +100,60 @@ std::string StreamUtils::GetCodecName(int codecId, int profile)
+
+std::string StreamUtils::GetLayout(uint64_t mask, int channels)
+{
+  if (mask)
```
```suggestion
  if (mask > 0)
```

CastagnaIT reviewed (commented):
i made a fast test by replacing StreamChannelsVar on Variables.xml with
```xml
	<variable name="StreamChannelsVar">
		<value condition="!String.IsEmpty(ListItem.Property(stream.channellayout))">$INFO[ListItem.Property(stream.channellayout)]</value>
		<value>$INFO[ListItem.Property(stream.channels)] channels</value>
	</variable>
```
seem to be ok
just to finish skin code changes

i have nothing else to say,
If in the next few days no one comments about the DB change to 
mask int  VS layout z.y.z string
I will approve without further question

the only thing is that its not very clear to me the use cases that might need the mask int for future improvements, an example would be appreciated

CrystalP commented:
I did the mask approach first because why not keep all the available details. But it has a small cpu cost at runtime, it's not as easy to work with in .nfo or media managers, and the need doesn't exist (will it ever?). 

Its value would come from the ability to build smart playlists or filter list of items that include or exclude certain speaker positions for example. Most of that could also be done with the channel count or the x.y.z form, like filter movies with surround sound or lfe.

At this time I'd say that using the x.y.z string throughout makes more sense. See the other PR #26747 with that approach.

CrystalP committed:
[VideoInfoTag] Add Import/export of channelmask .nfo tag

jjd-uk commented:
My initial format testing show that VideoPlayer.AudioChannelLayout doesn't seem to like 7.1 channels for some codec in a m2ts container.

ListItem.AudioChannelLayout looks to be ok
![image](https://github.com/user-attachments/assets/17b4b122-41a4-4af5-a4b4-9314bdea9b57)

VideoPlayer.AudioChannelLayout doesn't show any channels

![image](https://github.com/user-attachments/assets/f84ec51b-a4af-4dd0-9f2d-a0946ff21cf7)

Same across multiple codecs

![image](https://github.com/user-attachments/assets/21de612e-3d01-46f4-8810-1cd453e82fb0)

Here as well as missing channel layout incorrect codec is showing, DTS is shown when it should be DTS-HD MA

![image](https://github.com/user-attachments/assets/aa8a774d-f0f8-42ef-82a4-7fe29c1587d5)

TrueHD 7.1 appears to be ok though

![image](https://github.com/user-attachments/assets/8d58f88e-5641-4459-9ac1-939d794f01da)

![image](https://github.com/user-attachments/assets/322b9f5d-e1be-4040-a6d3-a7335683e33c)

CrystalP commented:
m2ts: most likely not related to the PR, m2ts and ts often report incomplete stream information while playing due to double-opening that's supposed to help with fast channel switching I think, but information is lost in the process.
You should notice the same thing in master. I hit something similar while testing the earlier 26618

CrystalP committed:
[VideoPlayer] Modify codecdesc to always contain a codec and channel part

The skin doesn't have to display x channels or similar after codecdesc "just in case" (ex. Estuary OSD audio settings)

jjd-uk commented:
> m2ts: most likely not related to the PR, m2ts and ts often report incomplete stream information while playing due to double-opening that's supposed to help with fast channel switching I think, but information is lost in the process. You should notice the same thing in master. I hit something similar while testing the earlier 26618

It looks like you're right it's the same with Master, I hadn't noticed that before. However what has this got to do with channel switching? since as far as I'm aware m2ts is exclusively used as the container type for Blu-Ray.

CrystalP committed:
to merge with commit 3 - don't depend on ffmpeg for default masks.

CrystalP commented:
in the last force-push I changed things so that only known masks are persisted in the db and existing data, nfo without the new tag and inputstream get the default mask at runtime, so no observable changes for the user compared to previous iterations of the PR. I thought it would help with migration of existing data, but it's not that helpful. I'm still searching.

The additional commit provides Kodi code to find a default mask from a number of channels, instead of calling into ffmeg with some overrides. No user impact for this one either, but simplifies the code I believe. The binary search could easily be replaced with a simple linear search if others feel it's overkill.
